ibtk/include/ibtk/HierarchyAveragedDataManager.h

```diff
@@ -0,0 +1,293 @@
+     */
+    inline const std::set<double>& getSnapshotTimePts()
+    {
+        return d_snapshot_time_pts;
+    }
+
+    /*!
+     * Returns the exact data point at which points are being stored in the internal map. Note that if a time point is
+     * not found within the specified tolerance, an unrecoverable error occurs.
+     */
+    double getTimePt(double time, double tol);
+
+    /*!
+     * Get the SnapshotCache object.
+     */
+    SnapshotCache& getSnapshotCache()
+    {
+        return d_snapshot_cache;
+    }
+
+    /*!
+     * Set the threshold for achieving a steady state. This class determines whether a periodic steady state has been
+     * achieved by checking the whether \f$ \frac{1}{N}||u - u_avg||_2 \f$ is less than the provided threshold.
+     */
+    void setSteadyStateThreshold(const double threshold)
+    {
+        d_periodic_thresh = threshold;
+    }
+
+private:
+    /*!
+     * \brief Registers a scratch variable with the variable database. This data is allocated and deallocated as needed.
+     */
+    void commonConstructor(SAMRAI::tbox::Pointer<SAMRAI::tbox::Database> input_db,
+                           SAMRAI::tbox::Pointer<SAMRAI::hier::PatchHierarchy<NDIM> > hierarchy);
+
+    std::string d_object_name;
+
+    /*
+     * Data for tracking mean flow quantities.
+     */
+    unsigned int d_num_averaging_cycles = 0;
+    SAMRAI::tbox::Pointer<SAMRAI::hier::Variable<NDIM> > d_var;
+    int d_scratch_idx = IBTK::invalid_index;
+    std::string d_mean_refine_type = "CONSERVATIVE_LINEAR_REFINE";
+
+    /*
+     * Interval to keep flow quantities
+     */
+    double d_t_start = std::numeric_limits<double>::quiet_NaN(), d_t_end = std::numeric_limits<double>::quiet_NaN(),
+           d_t_period = std::numeric_limits<double>::quiet_NaN(),
+           d_periodic_thresh = std::numeric_limits<double>::quiet_NaN();
+
+    /*
+     * Map between time point and the number of updates. We use this value to update average snapshots.
+     *
+     * Note we have to be careful when retrieving values from this map. Values should ONLY be retrieved and stored using
+     * the values stored in d_snapshot_time_pts.
+     */
+    std::map<double, unsigned int> d_idx_num_updates_map;
+
+    /*
+     * Map between time point and whether it's at a steady state.
+     *
+     * Note we have to be careful when retrieving values from this map. Values should ONLY be retrieved and stored using
+     * the values stored in d_snapshot_time_pts.
+     */
+    std::map<double, bool> d_idx_steady_state_map;
+
+    /*
+     * Set of time points that snapshots are determined. These values are between d_t_start and d_t_end.
+     */
+    std::set<double> d_snapshot_time_pts;
+
+    SnapshotCache d_snapshot_cache;
+
+    SAMRAI::tbox::Pointer<SAMRAI::math::HierarchyDataOpsReal<NDIM, double> > d_hier_data_ops;
+
+    // Drawing stuff
+    std::unique_ptr<SAMRAI::appu::VisItDataWriter<NDIM> > d_visit_data_writer;
```

Comment:
Bad news: this isn't going to to work in all cases because `VisItDataWriter` is a de facto singleton since it can only handle a single summary file (see `VisItDataWriter.C` circa line 3724). What happens to the original output when you write this? Does writing one summary file reset the other?

Reply:
I've done this several times for different projects and have never had a problem, as long as files are written to different directories. Granted, I haven't paid close attention to the details of what's happening. I'll take a closer look.

Reply:
As long as the `VisItDataWriter` objects write to different directories, there are no obvious problems. You get two summary files that are correct and no data is overwritten.

Reply:
Hm, I suppose this is OK. We could 'fix' it in ibsamrai2 and then require that everyone update their SAMRAI installation - I'll let you decide on whether or not we should do that.

Reply:
I added a warning to the documentation about writing to different directories.